File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/shared/promises/extensions/DeferredPromise.java
package com.lasthopesoftware.bluewater.shared.promises.extensions;

import com.namehillsoftware.handoff.promises.Promise;

public class DeferredPromise<Resolution> extends Promise<Resolution> {
	private final Resolution resolution;
	private final Throwable error;

	public DeferredPromise(Resolution resolution) {
		this.resolution = resolution;
		this.error = null;
	}

	public DeferredPromise(Throwable error) {
		this.resolution = null;
		this.error = error;
	}

	public void resolve() {
		if (this.resolution != null) resolve(resolution);
		else reject(error);
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/shared/promises/extensions/DeferredPromise.kt
package com.lasthopesoftware.bluewater.shared.promises.extensions

import com.namehillsoftware.handoff.cancellation.CancellationResponse
import com.namehillsoftware.handoff.promises.Promise

open class DeferredPromise<Resolution> : Promise<Resolution>, CancellationResponse {
    val resolution: Resolution?
    val error: Throwable?

	var isCancelled = false
		private set

    constructor(resolution: Resolution) {
        this.resolution = resolution
        error = null
    }

    constructor(error: Throwable?) {
        resolution = null
        this.error = error
    }

	init {
	    awaitCancellation(this)
	}

    fun resolve() {
        if (resolution != null) resolve(resolution)
		else reject(error)
    }

	override fun cancellationRequested() {
		isCancelled = true
	}
}
